dbms/src/Interpreters/ExpressionActions.cpp

```diff
@@ -990,6 +994,96 @@ void ExpressionActions::optimizeArrayJoin()
     }
 }
 
+void ExpressionActions::compileFunctions([[maybe_unused]] const Names & output_columns)
+{
+#if USE_EMBEDDED_COMPILER
+    LLVMContext context;
+    std::vector<bool> redundant(actions.size());
+    // an empty optional is a poisoned value prohibiting the column's producer from being removed
+    // (which it could be, if it was inlined into every dependent function).
+    std::unordered_map<std::string, std::unordered_set<std::optional<size_t>>> current_dependents;
```

Comment:
It's better use struct `{ bool is_used; std::unordered_set<size_t> dependents; }` instead of `std::unordered_set<std::optional<size_t>>`